### ISSUE TITLE
No `effect.flat_map`?

### ISSUE THREAD

**Issue by ghivert** [lustre-labs/lustre#75]
Hi!

While there's an `effect.map`, there's no `effect.flat_map`, which could be useful to handle effect chaining instead of having to rely on the `update` chain? 
We could need to cascade HTTP calls for example, and it would respect the usual Monad interface 🙂

I can understand it could be an "advanced" feature, but I think it could be nice to open possibilities for development, do you have any opinion on this?

**Comment by hayleigh-dot-dev** [comment]
The effect concept is lifted straight from elm, and in both cases you can think of them as a monoid (`effect.none` is mempty, `effect.batch` is mconcat) + functor. The thinking behind this is there is no _guarantee_ that an effect ever dispatches anything so providing a way to sequence effects may lead to confusing cases where some reasonable-looking code is never executed because an effect somewhere in the chain doesnt dispatch.

Elm "solves" this by having a separate type called a `Task` which roughly represents "the set of effects that are monadic". 

---

My general approach to lustre for things where i dont have strong opinions is "do what elm does until theres a reason to do otherwise" and so far no one has needed/wanted effect chaining. I think we have a few options:

1) do nothing and expect folks to chain messages themselves
2) ask the lustre_http author to provide some way of chaining http requests specifically
3) come up with our own task abstraction (and ask lustre_http to add task versions of their effects)
4) make the existing effect type monadic

All of these have some shortcomings in my opinion...

1) may end up in a bunch of intermediary message variants that dont do anything but thread efffects – this could be a good or a bad thing depending on your perspective
2) we'd end up asking each individual package for this, and there would be no way to chain effects from other libraries together
3) this is another concept for folks to learn, and the difference between "task" and "effect" is subtle enough to feel redundant
4) making the current effects monadic would mean having `effect.new: fn(a) -> Effect(a)` and i fear that might encourage folks to write effects that do nothing but dispatch a message: often a sign of poor message design and a misunderstanding of MVU.

Which is to say I'm inclined to sit on it for now and see if a better solution emerges (or more people complain). It would be trivial enough to make effects monadic but impossible to undo once we go down that path so I'd like to hold off for a while.

**Comment by ghivert** [comment]
Thanks for the detailed explanations. My last real elm project is a little old, and my brain sometimes forgot about things ha ha. By the way, the issue was mainly there to understand the rationales behind the design choice to avoid doing stuffs completely different (because in this case, it would probably be better to just fork lustre and go my own path, which is not something I want, I prefer to find a nice consensus and consolidate around one codebase and mutualize efforts). Maybe it could be moved to discussions 🙂

---

> My general approach to lustre for things where i dont have strong opinions is "do what elm does until theres a reason to do otherwise" and so far no one has needed/wanted effect chaining.

I deeply understand, and I think it's a good thing to stick with what works, especially since elm is working great and built a strong community around the tech, while lustre fixes almost everything that I personally found frustrating in the experience, so it seems a good thing. That being said, I understand the arguments for the different aspects.

I'm not a fan of 1., specifically because I often ended up having to use a bunch of intermediary messages when a lot of effects are needed, and it's something that I'm not really fan. I completely understand that some find it okay, and I don't want to "force" people to integrate chaining if that's not needed.

I agree with the rest of the arguments. I was more inclined to define something like task too, and I was about to write something like `EffectMonad` on my own. However, I'm not sure the distinction between `Effect` and `Task` are really clear for me in elm though. I feel like it's a somewhat arbitrary frontier between the two, because a `Task` like `process.sleep(10_000_000)` will probably never fire from your point of view, leading to a situation where the distinction between an `Effect` and a `Task` is somewhat blurry. 

> i fear that might encourage folks to write effects that do nothing but dispatch a message: often a sign of poor message design and a misunderstanding of MVU.

If I'm not wrong, I think you're talking about doing something like `effect.from(fn (dispatch) { dispatch(lustre.dispatch(MyMsg)) })`? But in the end, what is the difference between this and having a `effect.new` function? If that's the case, I think when someone want to do bad message design, that people will be able to ha ha 😅

---

All in all, I'll follow the `Task` route at the moment, it's not that much work, and it's plainly compatible with the `Effect` system. I also think it's the best thing to do right now because it's always possible to merge `Task` into `Effect` later while doing the reverse (splitting `Task` from monadic `Effect`) would be just impossible once integrated.

I'd be more than happy to share my experiments when I'll have something to show if you're interested in this 😉 (I'm working at the moment to go the full gleam stack route, with lustre as the main frontend framework)

**Comment by hayleigh-dot-dev** [comment]
> I feel like it's a somewhat arbitrary frontier between the two

Yeah I agree, and it's a semi-regular point of discussion in Elm too.

On the point about "if someone really wants to dispatch messages they can already" – it's true that if they _really_ want to they can, but the api is cumbersome enough that you aren't _encouraged_ to. Writing `effect.new(MyMsg)` would be a much more convenient incantation. What we make easy in the api is what people will do, so it's important the mindful of what things a new function makes easy and consider whether we want those things!

---

For time time-being you can actually hold off inventing your own task thing because you _can_ technically implement monadic chaining today with [`effect.perform`](https://hexdocs.pm/lustre/lustre/effect.html#perform). This is _technically_ an internal API that is only documented because the `@internal` attribute didn't exist at the time of publishing, so it'll disappear in the docs in the near future. I don't plan on touching it any time soon though, so you could make use of it and implement your own chaining

```gleam
fn then(effect: Effect(a), next: fn(a) -> Effect(b)) -> Effect(b) {
  // This is necessary because `perform` needs an implementation for
  // `effect.emit` but we can't handle that. If you're not using components
  // this will never come up.
  //
  let dummy_emit = fn(_, _) { Nil }
  
  use dispatch <- effect.from
  let run_next = fn(a) {
    effect.perform(next(a), dispatch, dummy_emit)
  }
  
  effect.perform(effect, run_next, dummy_emit)
}
```

---

**Comment by ghivert** [comment]
Thanks a lot for the help!

**Comment by spicychickensauce** [comment]
Hey @hayleigh-dot-dev . I found this issue, looking for something like `effect.then`.
I then tried to use your snippet for defining my own, but the signature of `effect.perform` has changed significantly, so I couldn't figure out how to adapt it to do that.

Is it still possible to define `then`? Do you still not want to include a `then` officially?

I have used Elm in the past and I always preferred using Task everywhere if possible. I think chaining just comes up quite often in practice.

My use case for wanting a `then`:
I'm storing a bunch of things in local storage (actually indexedDB but doesn't matter). I have created some effects for loading different values. Now I need a new effect that has to load a value from a different key, depending on the stored value in another key.
If I had `effect.then` I could just reuse my already defined effects and be done with it.
Without it, I either have to thread multiple updates (very bad!) or define a new custom effect that handles both in one (better, but repetitive).

**Comment by yoshi-monster** [comment]
Multiple updates aren't bad, you can even implement a generic `effect.then` on top of it
I wouldn't use `effect.perform` it requires much more bookkeeping on your end to implement correctly these days.
It's also a fully internal API now that has changed quite frequently over the course of v5.

**Comment by spicychickensauce** [comment]
@yoshi-monster I think multiple updates are bad because they hide control flow and make it hard to follow a simple fetch this, then that chain.
I guess I could implement an `effect.then` like thing via update using a message with a function in it, but this was discouraged in Elm days. Is it ok in lustre? Or were you thinking of another way to implement the generic `effect.then` via update?

**Comment by yoshi-monster** [comment]
The reasons for both of these things are the same, but I personally also feel the opposite about threading things through update :D
If you liked the `Task` versions in Elm the Gleam version of that is writing `Promise`-returning functions.

**Comment by spicychickensauce** [comment]
> The reasons for both of these things are the same, but I personally also feel the opposite about threading things through update :D

Could you elaborate? I don't understand what you meant by that?

Honestly I kinda forgot about promise... I guess I'll change my internal API to use promise, I think that should work fine for me. Thanks!

So is `effect.then` just not needed as we should use promise instead? But if that is the case, shouldn't e.g. rsvp use promises instead of effect? For me it's fine to switch to promises here, as I fully control this API, but if I wanted to mix it with some other code that uses effect I couldn't. So I think there is still a strong argument for `effect.then` to exist.

**Comment by hayleigh-dot-dev** [comment]
There's nothing wrong with producing and handling multiple messages and they explicitly do the opposite of hiding control flow because they give a name (the message) to each step in the chain: `effect.then` would be the one hiding control flow!

It's important to remember that an `Effect(a)` represents any number of side effects the same way `List(a)` could hold any number of elements. Because effects,, perform _side effects_ they are sensitive to the number of times they are run but you cannot know how many times an effect chained with `then` would run:

```gleam
effect.batch([wibble, wobble])
|> effect.then(woo)
```

`woo` would run twice here! It's not particularly egregious when you can see the `effect.batch` call right above it, but it would be easy to run into confusing scenarios as soon as then `.then` part is not right next to the original effect(s).

---

To look at other communities let's consider React and `useEffect`. If a library, let's a say a http fetching library, provides its own effects it would do so by exporting a hook like:

```js
export function useQuery(..) {
  useEffect(() => {
    fetch(..)
  })
}
```

Hooks don't compose in react so it would not be possible to write

```js
useEffect(() => {
  useQuery(..).then(useQuery(..))
})
```

So why isn't this a problem for the React community? Providing a hook is a way of providing an _opinionated api_ with a fixed shape: when i think fetching should be done a particular way i'd publish a package with my `useQuery` hook and have my users follow my patterns and design.

If i _don't_ want to constraint how people interact with my code then i wouldn't provide a hook, i'd provide the effectful functions directly and let them compose them however they like in their own `useEffect` call.

The same holds for Lustre. If a library is providing `Effect`s it's because its providing a specific interface to interact with some effectful thing(s), so that's what `rsvp` does for example it abstracts `gleam_fetch` in an opinionated way. For programs that don't like those opinions they would simply not use rsvp and either find another package that provides a different api or interface with the effectful functions directly: in this case by using `gleam_fetch` and chaining promises.

Compared to Elm Gleam is very good in this scenario because you can write both effectful code and "pure" lustre code side by side so you already can write the code that interacts with local storage exactly how you'd like it and then wrap it up in an `effect.from` call whenever.

---

i get the feeling that its the "wrap it up in an effect" that you are feeling friction in:

```gleam
wibble
|> effect.then(wobble)
```

vs

```gleam
wibble_and_wobble()

// elsewhere
fn wibble_and_wobble() {
  use _ <- effect.from

  wibble()
  |> wobble
}
```

but given effects are the way your program interacts with external systems i would say that giving them domain-appropriate names by abstracting them into function calls would be best practice _regardless of if `effect.then` existed_ so the comparison becomes:

```gleam
fn wibble_and_wobble() {
  wibble
  |> effect.then(wobble)
}
```

vs

```gleam
fn wibble_and_wobble() {
  use _ <- effect.from

  wibble()
  |> wobble
}
```

and once we're here i think the benefit of adding `effect.then` would be minimal indeed.

**Comment by spicychickensauce** [comment]
Thanks @hayleigh-dot-dev for taking the time to write this out, now I understand. Especially the problem with `batch` helped, thanks!

But hooks do compose, it's common to do a useQuery that depends on a previous useQuery, see e.g. https://tanstack.com/query/latest/docs/framework/react/guides/dependent-queries.
And hooks also fire multiple times, like that `effect.batch` example. (`useEffect` doesn't return anything, so it's not really a good comparison)
Of course, hooks also hold state and effects in lustre don't, so it's not quite the same.

So I'll be composing things outside effects, but I'm still not fully convinced that `effect.then` is a bad idea, especially given the comparison to hooks..
